flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/statistics/ContinuousHistogram.scala

```diff
@@ -0,0 +1,315 @@
+    data.insert(search, (p, 1))
+    // If we're over capacity or any two elements are within 1e-9 of each other, merge.
+    // This will take care of the case if p was actually equal to some value in the histogram and
+    // just increment the value there
+    mergeElements()
+  }
+
+  /** Merges the histogram with h and returns a histogram with capacity B
+    *
+    * @param h histogram to be merged
+    * @param B capacity of the resultant histogram
+    * @return Merged histogram with capacity B
+    */
+  override def merge(h: OnlineHistogram, B: Int): ContinuousHistogram = {
+    h match {
+      case temp: ContinuousHistogram => {
+        val m: Int = bins
+        val n: Int = temp.bins
+        var i, j: Int = 0
+        val mergeList = new mutable.ArrayBuffer[(Double, Int)]()
+        while (i < m || j < n) {
+          if (i >= m) {
+            mergeList += ((temp.getValue(j), temp.getCounter(j)))
+            j = j + 1
+          } else if (j >= n || getValue(i) <= temp.getValue(j)) {
+            mergeList += data.apply(i)
+            i = i + 1
+          } else {
+            mergeList += ((temp.getValue(j), temp.getCounter(j)))
+            j = j + 1
+          }
+        }
+        // the size will be brought to capacity while constructing the new histogram itself
+        val finalLower = Math.min(lower, temp.lower)
+        val finalUpper = Math.max(upper, temp.upper)
+        val ret = new ContinuousHistogram(B, finalLower, finalUpper)
+        ret.loadData(mergeList.toArray)
+        ret
+      }
+      case default =>
+        throw new RuntimeException("Only a continuous histogram is allowed to be merged with a " +
+          "continuous histogram")
+
+    }
+  }
+
+  /** Returns the qth quantile of the histogram
+    *
+    * @param q Quantile value in (0,1)
+    * @return Value at quantile q
+    */
+  def quantile(q: Double): Double = {
+    require(bins > 0, "Histogram is empty")
+    require(q > 0 && q < 1, "Quantile must be between 0 and 1")
+    val wantedSum = (q * total).round.toInt
+    var currSum = count(getValue(0))
+
+    if (wantedSum < currSum) {
+      require(lower > -MaxValue, "Set a lower bound before proceeding")
+      return Math.sqrt(2 * wantedSum * Math.pow(getValue(0) - lower, 2) / getCounter(0)) + lower
+    }
+
+    /** Walk the histogram to find sums at every bin value
+      * As soon as you hit the interval where you should be
+      * Walk along the trapezoidal line
+      * This leads to solving a quadratic equation.
+      */
+    for (i <- 1 to bins - 1) {
+      val tmpSum = count(getValue(i))
+      if (currSum <= wantedSum && wantedSum < tmpSum) {
+        val neededSum = wantedSum - currSum
+        val a: Double = getCounter(i) - getCounter(i - 1)
+        val b: Double = 2 * getCounter(i - 1)
+        val c: Double = -2 * neededSum
+        if (a == 0) {
+          return getValue(i - 1) + (getValue(i) - getValue(i - 1)) * (-c / b)
```

Comment:
Scala has a break statement. However, it doesn't let you use it that easily for a good reason. It's not good style because it leads to a complex control flow. In Scala `if` statements are in fact an expression and can return a value. Thus you can write `val x = if(...) 1 else 2`.